### PR TITLE
[SPARK-29545][SQL] Add support for bit_xor aggregate function

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

bit_xor(expr) - Returns the bitwise XOR of all non-null input values, or null if none

### Why are the changes needed?

As we support `bit_and`, `bit_or` now, we'd better support the related aggregate function **bit_xor** ahead of postgreSQL, because many other popular databases support it.

 

http://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.help.sqlanywhere.12.0.1/dbreference/bit-xor-function.html

https://dev.mysql.com/doc/refman/5.7/en/group-by-functions.html#function_bit-or

https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/Aggregate/BIT_XOR.htm?TocPath=SQL%20Reference%20Manual%7CSQL%20Functions%7CAggregate%20Functions%7C_____10

 
### Does this PR introduce any user-facing change?

add a new bit agg
### How was this patch tested?

UTs added

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/postgreSQL/aggregates_part2.sql.out [hunk 1]
```diff
@@ -191,113 +206,113 @@ SELECT
...
+select max(unique2) from tenk1 order by max(unique2)
 -- !query 26 schema
 struct<max(unique2):int>
 -- !query 26 output
 9999
 
 
 -- !query 27
-select t1.max_unique2, g from (select max(unique2) as max_unique2 FROM tenk1) t1 LATERAL VIEW explode(array(1,2,3)) t2 AS g order by g desc
+select max(unique2) from tenk1 order by max(unique2)+1
```
Comment (srowen) [inline sql/core/src/test/resources/sql-tests/results/postgreSQL/aggregates_part2.sql.out]: Just to be clear, these aren't real differences, just how github is interpreting the diff?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala [hunk 2]
```diff
@@ -97,3 +97,42 @@ case class BitOrAgg(child: Expression) extends DeclarativeAggregate with Expects
 
   override lazy val evaluateExpression: AttributeReference = bitOr
 }
+
+@ExpressionDescription(
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala]: This is a dumb question, but, where is the actual bitwise-xor implemented? I just can't see it. For my own info.

**Review by srowen — COMMENTED** [review]
I dunno, @cloud-fan what do you think of adding this? seems niche but reasonable.

**Comment by srowen** [comment]
Also do you need to update docs or Pyspark with this function?

File: sql/core/src/test/resources/sql-tests/results/postgreSQL/aggregates_part2.sql.out (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/results/postgreSQL/aggregates_part2.sql.out]: there are not real diff

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala]: We can implement `updateExpressions`,  `mergeExpressions ` logically, and they would be eventually codegened in `*Exec`s like `HashAggregateExec`

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala]: @srowen we can treat `Expression` as a DSL, then this function is implemented by the DSL.

**Comment by cloud-fan** [comment]
> Also do you need to update docs or Pyspark with this function?

The newly added bitwise functions are only available in SQL.

This is a general question: shall we add scala/python/R functions for all the builtin SQL functions? I feel we only need the common ones, as users can always do `expr("sql_fun(para1)")` to access SQL functions in scala/python/R.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala [hunk 3]
```diff
@@ -97,3 +97,42 @@ case class BitOrAgg(child: Expression) extends DeclarativeAggregate with Expects
...
+  usage = "_FUNC_(expr) - Returns the bitwise XOR of all non-null input values, or null if none.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(col) FROM VALUES (3), (5) AS tab(col);
+       6
+  """,
+  since = "3.0.0")
+case class BitXorAgg(child: Expression) extends DeclarativeAggregate with ExpectsInputTypes {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala]: `BitAndAgg`, `BitOrAgg`, and `BitXorAgg` has the similar logic, so can we share it by using a new trait (e.g., BitwiseOpLike)?

**Comment by yaooqinn** [comment]
@srowen I guess docs will be generated via `@ExpressionDescription` annotation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala]: sounds good, will update

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala]: I think you can make it much simpler like this;
```
abstract class BitAggregate extends DeclarativeAggregate with ExpectsInputTypes {

  ...
  override lazy val updateExpressions: Seq[Expression] =
    If(IsNull(bitAgg),
      child,
      If(IsNull(child), bitAgg, bitOp(bitAgg, child))) :: Nil

  override lazy val mergeExpressions: Seq[Expression] =
    If(IsNull(bitAgg.left),
      bitAgg.right,
      If(IsNull(bitAgg.right), bitAgg.left, bitOp(bitAgg.left, bitAgg.right))) :: Nil
}

case class BitAndAgg(child: Expression) extends BitAggregate {

  override def nodeName: String = "bit_and"
  override def bitOp(left: Expression, right: Expression): BinaryArithmetic =
    BitwiseAnd(left, right)
}
```

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/bitwiseAggregates.scala]: yea, much cooler, thanks

**Comment by maropu** [comment]
The tests in `sql-tests/inputs/postgreSQL` have been just ported from the pgSQL repo, so can you add tests in anothe place, e.g.,  `sql-tests/inputs/bitwise.sql`?

**Comment by yaooqinn** [comment]
just for curiosity, if add same test in `sql-tests/inputs/bitwise.sql`, these tests will exist in three places including `sql-tests/inputs/postgreSQL/udf-aggregates_part2.sql` and `sql-tests/inputs/postgreSQL/aggregates_part2.sql`, are they too much redundant?

**Comment by maropu** [comment]
I meant the tests needs to be moved from `postgreSQL/*` to `bitwise.sql`.

**Comment by yaooqinn** [comment]
ok i got it

**Comment by maropu** [comment]
retest this please

**Review by maropu — APPROVED** [review]
LGTM @cloud-fan @srowen can you recheck?

**Review by srowen — APPROVED** [review]
Still looks reasonable to me.

**Comment by maropu** [comment]
Thanks, @srowen. Merged to master.

**Pull request closed without merging**